WordPress/gutenberg
# [Mobile] Adds useBlockDropZone and DroppingInsertionPoint

Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4710

## What?
This PRs adds the native implementation of `useBlockDropZone` as well as a new component `DroppingInsertionPoint`.

## Why?
This is part of the Drag & Drops Blocks project.

## How?

### `useBlockDropZone`

It pretty much reuses the same code as web, with a few changes to make it compatible with mobile:

- `getNearestBlockIndex`: receives the blocksLayouts object instead of DOM elements, as well as manually creating the `rect` element (which just has the size and position of the block in the screen) instead of using `getBoundingClientRect`.
- `useBlockDropZone`: 
  - `targetBlockIndex` is stored in a `SharedValue` instead of the state, this is to make it more performant.
  - It uses the new `getBlockLayoutsOrderedByYCoord` which returns the block's layouts sorted by their Y coordinate.
  - It returns an object with the handlers `onBlockDragOver`, `onBlockDragEnd` and the shared value `targetBlockIndex`.

### `DroppingInsertionPoint`

It's a new component that will show where a block can be dropped, on the web editor the `InsertionPoint` component handles this among other cases. Since our case it's a bit different I separated this into its own component.

The first approach I took was to reuse the same implementation as web that's connected to redux by using the `getBlockInsertionPoint` selector. This unfortunately causes a bad performance experience, especially on Android. So I changed it to use a shared value that's coming from the `useBlockDropZone` hook: `targetBlockIndex`.

So whenever the value of `targetBlockIndex` changes, it'll trigger `setIndicatorPosition` which will calculate where the indicator should be shown as the block is being dragged.

Kudos to @fluiddot for helping me with a bug in https://github.com/WordPress/gutenberg/pull/39891/commits/42c419a1e5cbe0819ba52f1966fb51da1a0898ed 🎉 

## Testing Instructions
- Open the app.
- Add several blocks or use the `initialHTML` content.
- Long-press on a block to activate the dragging mode.
- Move the block around and **expect** to see a blue horizontal line indicator.
- **Expect** to see the line either above or below the block you're dragging over.
- **Expect** to be able to move the block to the very top, the indicator **should** show up above the first block.
- **Expect** to be able to move the block to the very bottom, the indicator **should** show up below the last block.

## Screenshots or screencast 

## iOS

https://user-images.githubusercontent.com/4885740/161035162-03c0fad4-374f-4657-8864-3537799bfdaf.mov

## Android

https://user-images.githubusercontent.com/4885740/161035488-12195f45-d846-4e37-a1da-05b575be01f9.mov
